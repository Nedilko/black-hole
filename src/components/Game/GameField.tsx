import Cell from './Cell';
import type { IBoardSize } from '../../game/board';
import { useGameBoard } from '../../hooks/useGameBoard';

type PropsType = {
  size: IBoardSize;
  holesCount: number;
  onFinish: () => void;
  onOpenCell: () => void;
};

const GameField = ({ size, holesCount, onFinish, onOpenCell }: PropsType) => {
  const handleOpenCell = () => {
    onOpenCell();
  };
  const [board, remainingCellsCount] = useGameBoard(
    size,
    holesCount,
    onFinish,
    handleOpenCell
  );
  const { width, height } = size;

  return (
    <div className="flex flex-col">
      <div className="flex text-white hover:underline">
        {remainingCellsCount}
      </div>
      <div className="flex flex-col">
        <div
          className={`grid gap-2 grid-cols-${width} grid-rows-${height} p-4`}
        >
          {Array.from({ length: width * height }, (_, i) => {
            const { isHole, holesNearCount, isOpen, handleOpen } =
              board.cells[i];
            return (
              <Cell
                key={i}
                isOpen={isOpen}
                isHole={isHole}
                holesNearCount={holesNearCount}
                handleOpen={handleOpen}
              />
            );
          })}
        </div>
      </div>
    </div>
  );
};;

export default GameField;
