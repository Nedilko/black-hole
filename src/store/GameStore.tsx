import React, { useReducer, ReactNode } from 'react';

export interface IGameData {
  showGameField: boolean;
  showGameControls: boolean;
  remainingCellsCount: number;
  totalCellsCount: number;
  time: string;
  showTime: boolean;
}

const gameInitialSettings: IGameData = {
  showGameField: false,
  showGameControls: false,
  remainingCellsCount: 0,
  totalCellsCount: 0,
  time: '00:00',
  showTime: false,
};

const SHOW_GAME_FIELD = 'showGameField';
const HIDE_GAME_FIELD = 'hideGameField';
const SHOW_GAME_CONTROLS = 'showGameControls';
const HIDE_GAME_CONTROLS = 'hideGameControls';
const UPDATE_REMAINIG_CELLS_COUNT = 'updateRemainingCellsCount';
const UPDATE_TOTAL_CELLS_COUNT = 'updateTotalCellsCount';
const SHOW_TIME = 'showTime';
const HIDE_TIME = 'hideTime';

interface PayloadedAction<T, P> {
  type: T;
  payload: P;
}

interface Action<T> {
  type: T;
}

interface IShowGameFieldAction extends Action<typeof SHOW_GAME_FIELD> {}
interface IHideGameFieldAction extends Action<typeof HIDE_GAME_FIELD> {}
interface IShowGameControlsAction extends Action<typeof SHOW_GAME_CONTROLS> {}
interface IHideGameControlsAction extends Action<typeof HIDE_GAME_CONTROLS> {}
interface IUpdateRemainingCellsCountAction
  extends PayloadedAction<typeof UPDATE_REMAINIG_CELLS_COUNT, number> {}
interface IUpdateTotalCellsCountAction
  extends PayloadedAction<typeof UPDATE_TOTAL_CELLS_COUNT, number> {}
interface IShowTimeAction extends Action<typeof SHOW_TIME> {}
interface IHideTimeAction extends Action<typeof HIDE_TIME> {}

type GameActionsType =
  | IShowGameFieldAction
  | IHideGameFieldAction
  | IShowGameControlsAction
  | IHideGameControlsAction
  | IUpdateRemainingCellsCountAction
  | IUpdateTotalCellsCountAction
  | IShowTimeAction
  | IHideTimeAction;

function createPayloadedAction<
  T extends PayloadedAction<T['type'], T['payload']>
>(
  type: T['type']
): (payload: T['payload']) => PayloadedAction<T['type'], T['payload']> {
  return (payload: T['payload']) => ({
    type: type,
    payload,
  });
}

function createAction<T extends Action<T['type']>>(
  type: T['type']
): () => Action<T['type']> {
  return () => ({
    type: type,
  });
}

export const GameActions = {
  showGameField: createAction<IShowGameFieldAction>(SHOW_GAME_FIELD),
  hideGameField: createAction<IHideGameFieldAction>(HIDE_GAME_FIELD),
  showGameControls: createAction<IShowGameControlsAction>(SHOW_GAME_CONTROLS),
  hideGameControls: createAction<IHideGameControlsAction>(HIDE_GAME_CONTROLS),
  updateRemainingCellsCount:
    createPayloadedAction<IUpdateRemainingCellsCountAction>(
      UPDATE_REMAINIG_CELLS_COUNT
    ),
  updateTotalCellsCount: createPayloadedAction<IUpdateTotalCellsCountAction>(
    UPDATE_TOTAL_CELLS_COUNT
  ),
  showTime: createAction<IShowTimeAction>(SHOW_TIME),
  hideTime: createAction<IHideTimeAction>(HIDE_TIME),
};

function assertNever(never: never) {
  console.error('Never touched');
}

const settingsReducer = (state: IGameData, action: GameActionsType) => {
  switch (action.type) {
    case SHOW_GAME_FIELD: {
      return { ...state, showGameField: true };
    }
    case HIDE_GAME_FIELD: {
      return { ...state, showGameField: false };
    }
    case SHOW_GAME_CONTROLS: {
      return { ...state, showGameControls: true };
    }
    case HIDE_GAME_CONTROLS: {
      return { ...state, showGameControls: false };
    }
    case UPDATE_REMAINIG_CELLS_COUNT: {
      return { ...state, remainingCellsCount: action.payload };
    }
    case UPDATE_TOTAL_CELLS_COUNT: {
      return { ...state, totalCellsCount: action.payload };
    }
    case SHOW_TIME: {
      return { ...state, showTime: true };
    }
    case HIDE_TIME: {
      return { ...state, showTime: false };
    }
    default:
      assertNever(action);
      return state;
  }
};

type ProviderProps = {
  children: ReactNode;
};

export type Dispatch = (action: GameActionsType) => void;

const GameDataContext = React.createContext<IGameData | undefined>(undefined);
const GameDispatchContext = React.createContext<Dispatch | undefined>(
  undefined
);

const GameContextProvider = ({ children }: ProviderProps) => {
  const [state, dispatch] = useReducer(settingsReducer, gameInitialSettings);
  return (
    <GameDataContext.Provider value={state}>
      <GameDispatchContext.Provider value={dispatch}>
        {children}
      </GameDispatchContext.Provider>
    </GameDataContext.Provider>
  );
};

export { GameContextProvider, GameDataContext, GameDispatchContext };
