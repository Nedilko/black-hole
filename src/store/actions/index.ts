import { gameActions } from './game';
import { settingsActions } from './settings';

export { gameActions, settingsActions };
