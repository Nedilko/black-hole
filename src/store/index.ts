import {
  configureStore,
  combineReducers,
  PreloadedState,
} from '@reduxjs/toolkit';
import gameReducer from './gameSlice';
import fieldReducer from './fieldSlice';

const rootReducer = combineReducers({
  game: gameReducer,
  field: fieldReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
