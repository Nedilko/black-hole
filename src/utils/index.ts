import { getTime } from './time';
import {
  readFromLocalstorage,
  writeToLocalstorage,
} from './localstorageAdapter';
import { getDefaultResults } from './game';
import { loadResults, writeResults } from './dataAdapter';

export {
  getTime,
  readFromLocalstorage,
  writeToLocalstorage,
  getDefaultResults,
  loadResults,
  writeResults,
};
