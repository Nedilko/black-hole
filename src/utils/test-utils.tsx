import { RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from 'redux';
import { AppStore, RootState, setupStore } from '../store';
import { render } from '@testing-library/react';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  initialState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    initialState = {},
    store = setupStore(initialState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
