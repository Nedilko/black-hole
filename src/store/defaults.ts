import { SettingsState } from '../store/actions/settings';

const easySettings: SettingsState = {
  size: {
    width: 8,
    height: 8,
  },
  holesCount: 10,
};

const mediumSettings: SettingsState = {
  size: {
    width: 15,
    height: 15,
  },
  holesCount: 20,
};

const hardSettings: SettingsState = {
  size: {
    width: 20,
    height: 20,
  },
  holesCount: 30,
};

export const getDefaultSettings = (): SettingsState =>
  Object.create({
    size: Object.create(easySettings.size),
    holesCount: easySettings.holesCount,
  });

export const getSettings = (settings: SettingsState): SettingsState =>
  Object.create({
    size: Object.create(settings.size),
    holesCount: settings.holesCount,
  });

export const settings = {
  easy: getSettings(easySettings),
  medium: getSettings(mediumSettings),
  hard: getSettings(hardSettings),
};
