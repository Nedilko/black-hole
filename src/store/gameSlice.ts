import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import type { FieldSize } from './fieldSlice';
import type { AppDispatch } from '.';
import { renewField, setupField } from './fieldSlice';
import { loadResults } from '../utils';

export enum DIFFUCULTY {
  EASY,
  MEDIUM,
  HARD,
  CUSTOM,
}

export interface Results {
  [DIFFUCULTY.EASY]: number;
  [DIFFUCULTY.MEDIUM]: number;
  [DIFFUCULTY.HARD]: number;
  [DIFFUCULTY.CUSTOM]: number;
}

export interface GameState {
  showGameField: boolean;
  showGameControls: boolean;
  remainingCellsCount: number;
  totalCellsCount: number;
  showTime: boolean;
  showCounter: boolean;
  showSettings: boolean;
  time: number;
  isStarted: boolean;
  isFinished: boolean;
  isWon: boolean;
  difficulty: DIFFUCULTY;
  results: Results;
}

const defaultStateValue: GameState = {
  showGameField: false,
  showGameControls: false,
  remainingCellsCount: 0,
  totalCellsCount: 0,
  showTime: false,
  showCounter: false,
  showSettings: true,
  time: 0,
  isStarted: false,
  isFinished: false,
  isWon: false,
  difficulty: DIFFUCULTY.EASY,
  results: loadResults(),
};

const gameSlice = createSlice({
  name: 'game',
  initialState: defaultStateValue,
  reducers: {
    finishGame(state) {
      state.showGameControls = true;
      state.isFinished = true;
      state.isStarted = false;
    },
    showMainMenu(state) {
      state.showGameField = false;
      state.showGameControls = false;
      state.showCounter = false;
      state.showTime = false;
      state.showSettings = true;
      state.time = 0;
      state.isFinished = false;
      state.isWon = false;
    },
    showGameField(state) {
      state.showGameControls = false;
      state.showCounter = true;
      state.showTime = true;
      state.showSettings = false;
      state.showGameField = true;
      state.time = 0;
      state.isFinished = false;
      state.isWon = false;
    },
    updateTime(state) {
      state.time++;
    },
    startCountdown(state) {
      state.isStarted = true;
    },
    winGame(state) {
      state.isWon = true;
      state.showGameField = false;
      state.showGameControls = true;
      state.isFinished = true;
      state.isStarted = false;
      if (
        state.results[state.difficulty] > state.time ||
        state.results[state.difficulty] === 0
      ) {
        state.results[state.difficulty] = state.time;
      }
    },
    registerDifficulty(state, action: PayloadAction<DIFFUCULTY>) {
      state.difficulty = action.payload;
    },
  },
});

export default gameSlice.reducer;

export const {
  finishGame,
  showMainMenu,
  updateTime,
  winGame,
  registerDifficulty,
  startCountdown,
} = gameSlice.actions;

export const restartGame = () => (dispatch: AppDispatch) => {
  dispatch(renewField());
  dispatch(gameSlice.actions.showGameField());
};

export const startGame =
  (size: FieldSize, holesCount: number) => (dispatch: AppDispatch) => {
    dispatch(setupField({ size, holesCount }));
    dispatch(gameSlice.actions.showGameField());
  };
