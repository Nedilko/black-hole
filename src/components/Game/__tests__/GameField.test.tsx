import { setupStore } from '../../../store';
import { startGame } from '../../../store/gameSlice';
import { renderWithProviders } from '../../../utils/test-utils';
import GameField from '../GameField';
jest.mock('../Cell', () => () => <div>Cell</div>);

describe('GameField', () => {
  it('should render GameField with predefined width and height', () => {
    const store = setupStore();
    store.dispatch(startGame({ width: 10, height: 10 }, 1));
    const { container } = renderWithProviders(<GameField />, { store });
    expect(container).toMatchSnapshot();
  });
});
