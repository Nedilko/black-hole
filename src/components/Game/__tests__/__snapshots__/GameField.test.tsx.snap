// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField should render GameField 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="grid gap-2 grid-cols-10 grid-rows-10 p-4"
      >
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`GameField should render GameField with predefined width and height 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="grid gap-2 grid-cols-10 grid-rows-10 p-4"
      >
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
        <div>
          Cell
        </div>
      </div>
    </div>
  </div>
</div>
`;
